utils/s2n_key_material.c

```diff
@@ -0,0 +1,140 @@
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_key_material_update_data_offsets(struct s2n_key_material *key_material)
+{
+    RESULT_ENSURE_REF(key_material);
+
+    struct s2n_stuffer key_material_stuffer = { 0 };
+    struct s2n_blob key_material_blob = { 0 };
+    RESULT_GUARD_POSIX(s2n_blob_init(&key_material_blob, key_material->data, s2n_array_len(key_material->data)));
+    RESULT_GUARD_POSIX(s2n_stuffer_init_written(&key_material_stuffer, &key_material_blob));
+
+    /* initialize key_material blobs; incrementing ptr to point to the next slice of memory */
+    uint8_t *ptr = NULL;
+    /* MAC */
+    ptr = s2n_stuffer_raw_read(&key_material_stuffer, key_material->client_mac.size);
+    key_material->client_mac.data = ptr;
+    ptr = s2n_stuffer_raw_read(&key_material_stuffer, key_material->server_mac.size);
+    key_material->server_mac.data = ptr;
+
+    /* KEY */
+    ptr = s2n_stuffer_raw_read(&key_material_stuffer, key_material->client_key.size);
+    key_material->client_key.data = ptr;
+    ptr = s2n_stuffer_raw_read(&key_material_stuffer, key_material->server_key.size);
+    key_material->server_key.data = ptr;
+
+    /* IV */
+    ptr = s2n_stuffer_raw_read(&key_material_stuffer, key_material->client_iv.size);
+    key_material->client_iv.data = ptr;
+    ptr = s2n_stuffer_raw_read(&key_material_stuffer, key_material->server_iv.size);
+    key_material->server_iv.data = ptr;
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_key_material_calculate_size(struct s2n_key_material *key_material, struct s2n_connection *conn)
+{
+    RESULT_ENSURE_REF(key_material);
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(conn->secure);
+    RESULT_ENSURE_REF(conn->secure->cipher_suite);
+    RESULT_ENSURE_REF(conn->secure->cipher_suite->record_alg);
+    const struct s2n_cipher *cipher = conn->secure->cipher_suite->record_alg->cipher;
+    RESULT_ENSURE_REF(cipher);
+
+    uint8_t mac_size;
+    ssize_t key_size, iv_size = 0;
+
+    /* MAC size */
+    if (cipher->type == S2N_COMPOSITE) {
+        mac_size = cipher->io.comp.mac_key_size;
+    } else {
+        RESULT_GUARD_POSIX(s2n_hmac_digest_size(conn->secure->cipher_suite->record_alg->hmac_alg, &mac_size));
+    }
+
+    /* KEY size */
+    key_size = cipher->key_material_size;
+
+    /* IV size */
+    switch (cipher->type) {
+        case S2N_AEAD:
+            iv_size = cipher->io.aead.fixed_iv_size;
+            break;
+        case S2N_CBC:
+            iv_size = cipher->io.cbc.block_size;
+            break;
+        case S2N_COMPOSITE:
+            iv_size = cipher->io.comp.block_size;
+            break;
+        /* No-op for stream ciphers */
+        default:
+            break;
+    }
+
+    key_material->client_mac.size = mac_size;
+    key_material->server_mac.size = mac_size;
+    key_material->client_key.size = key_size;
+    key_material->server_key.size = key_size;
+    key_material->client_iv.size = iv_size;
+    key_material->server_iv.size = iv_size;
```

Comment:
This technically works, but is a strange way to treat blobs. It's likely to cause mistakes, since you now have blobs with NULL data but non-zero size.

Just calculate the sizes when you do the reads.